Update Microsoft.Identity.Model to 7.0.0-preview3

jennyf19 opened the pull request:
# {PR title}

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [ ] You've included unit or integration tests for your change, where applicable.
- [ ] You've included inline docs for your change, where applicable.
- [ ] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Summary of the changes (Less than 80 chars)

## Description

Update M.IM to 7.0.0-preview3, which fully  replaces Newtonsoft.Json with System.Text.Json, and enables ASP.NET's JwtBearer auth handler will now be fully AOT compatible. See [release notes](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/releases/tag/7.0.0-preview3) for details.



Fixes #{bug number} (in this specific format)

jennyf19 committed:
Update Microsoft.Identity.Model to 7.0.0-preview3

eerhardt commented:
/azp run

jennyf19 committed:
fixing null refs in WsFed

eerhardt commented on src/Security/Authentication/WsFederation/src/WsFederationHandler.cs:
```diff
@@ -98,7 +98,7 @@ protected override async Task HandleChallengeAsync(AuthenticationProperties prop
         {
             IssuerAddress = _configuration.TokenEndpoint ?? string.Empty,
-            Wtrealm = Options.Wtrealm,
+            Wtrealm = Options.Wtrealm!,
```
Is this the right fix? It seems wrong.

eerhardt replied:
It looks like the change that broke this is

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/commit/c058737642f6e04aab80fef5f15de118e7bb2adc

@brentschmaltz @westin-m @FuPingFranco - are you certain that `WsFederationMessage.Wtrealm` needs to be non-nullable?

brentschmaltz replied:
@eerhardt When requesting a SecurityToken, wtrealm must contain a uri.
http://[docs.oasis-open.org/wsfed/federation/v1.2/os/ws-federation-1.2-spec-os.html#_Toc223175023](http://docs.oasis-open.org/wsfed/federation/v1.2/os/ws-federation-1.2-spec-os.html#_Toc223175023)

eerhardt replied:
@halter73 @Tratcher - any thoughts here? Why is `Options.Wtrealm` annotated as `string?` when `WsFederationMessage.Wtrealm` is a non-nullable `string`. Do we have our nullable annotations incorrect in ASP.NET?

jennyf19 replied:
@eerhardt will you fix it on your end?

eerhardt replied:
I've opened https://github.com/dotnet/aspnetcore/issues/50242 to track annotating `Wtrealm` correctly, and added links to those 2 locations in the code. I believe the `Wctx` below is a bug in IdentityModel.

eerhardt commented on src/Security/Authentication/WsFederation/src/WsFederationHandler.cs:
```diff
@@ -195,7 +195,7 @@ protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync
                 // Extract the user state from properties and reset.
                 properties.Items.TryGetValue(WsFederationDefaults.UserstatePropertiesKey, out var userState);
-                wsFederationMessage.Wctx = userState;
+                wsFederationMessage.Wctx = userState!;
```
This looks like a bug in IdentityModel. `wctx` is OPTIONAL according to http://docs.oasis-open.org/wsfed/federation/v1.2/os/ws-federation-1.2-spec-os.html#_Toc223175023.

@jennyf19 @brentschmaltz - can you open an issue in https://github.com/azuread/azure-activedirectory-identitymodel-extensions-for-dotnet/ to track annotating these properties correctly for `nullable`? Then link to the bug here.

jennyf19 replied:
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2240

eerhardt committed:
Add link to tracking issue.

eerhardt committed:
Add link to tracking issue.

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/source-build-externals build 20230824.1

Microsoft.SourceBuild.Intermediate.source-build-externals
 From Version 8.0.0-alpha.1.23421.1 -> To Version 8.0.0-alpha.1.23424.1

eerhardt committed:
Merge branch 'release/8.0' into jennyf/M.IM.Preview3

eerhardt reviewed (approved):
LGTM.

@jennyf19 - Since we merged https://github.com/dotnet/source-build-externals/pull/194, we will need to take preview3 in order to unblock the build. When preview4 comes, we will update to that as well.

eerhardt committed:
Workaround https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2261 in a test.
